moshi/src/main/java/com/squareup/moshi/JsonValueWriter.java

```diff
@@ -182,6 +183,28 @@ public Object root() {
     return this;
   }
 
+  @Override public JsonWriter value(BufferedSource source) throws IOException {
+    Object value = JsonReader.of(source).readJsonValue();
+    if (value instanceof String) {
+      return value((String) value);
+    }
+    if (value instanceof Number) {
+      return value((Number) value);
+    }
+    if (promoteValueToName) {
+      throw new AssertionError("This is not a map key.");
+    }
+    boolean serializeNulls = this.serializeNulls;
+    this.serializeNulls = true;
```

Comment:
nice

Reply:
I might document this explicitly in the Javadoc.